substrate/frame/revive/src/lib.rs

```diff
@@ -620,50 +610,36 @@ pub mod pallet {
 				.checked_mul(2)
 				.expect("MaxTransientStorageSize is too large");
 
-			// Check that given configured `MaxCodeLen`, runtime heap memory limit can't be broken.
+			// Check that the configured `STATIC_MEMORY_BYTES` fits into runtime memory.
 			//
-			// In worst case, the decoded Wasm contract code would be `x16` times larger than the
-			// encoded one. This is because even a single-byte wasm instruction has 16-byte size in
-			// wasmi. This gives us `MaxCodeLen*16` safety margin.
-			//
-			// Next, the pallet keeps the Wasm blob for each
-			// contract, hence we add up `MaxCodeLen` to the safety margin.
+			// `STATIC_MEMORY_BYTES` is the amount of memory that a contract can consume
+			// in memory and is enforced at upload time.
 			//
 			// The inefficiencies of the freeing-bump allocator
 			// being used in the client for the runtime memory allocations, could lead to possible
-			// memory allocations for contract code grow up to `x4` times in some extreme cases,
-			// which gives us total multiplier of `17*4` for `MaxCodeLen`.
+			// memory allocations grow up to `x4` times in some extreme cases.
 			//
-			// That being said, for every contract executed in runtime, at least `MaxCodeLen*17*4`
-			// memory should be available. Note that maximum allowed heap memory and stack size per
-			// each contract (stack frame) should also be counted.
+			// Dynamic allocations are metered at runtime as soon as they become available and are
+			// not part of this calculation.
 			//
 			// The pallet holds transient storage with a size up to `max_transient_storage_size`.
 			//
 			// Finally, we allow 50% of the runtime memory to be utilized by the contracts call
 			// stack, keeping the rest for other facilities, such as PoV, etc.
-			//
-			// This gives us the following formula:
-			//
-			// `(MaxCodeLen * 17 * 4 + MAX_STACK_SIZE + max_heap_size) * max_call_depth +
-			// max_transient_storage_size < max_runtime_mem/2`
-			//
-			// Hence the upper limit for the `MaxCodeLen` can be defined as follows:
-			let code_len_limit = max_runtime_mem
+			let static_memory_limit = max_runtime_mem
 				.saturating_div(2)
 				.saturating_sub(max_transient_storage_size)
 				.saturating_div(max_call_depth)
-				.saturating_sub(max_heap_size)
-				.saturating_sub(MAX_STACK_SIZE)
-				.saturating_div(17 * 4);
+				.saturating_sub(STATIC_MEMORY_BYTES)
+				.saturating_div(4);
```

Comment:
```suggestion
				.saturating_div(EXTRA_OVERHEAD_PER_CODE_BYTE);
```

Reply:
This is to account for memory allocator overhead (see the docs above). But it's a good point: All those numbers should have names.

Reply:
Replaced by constants